### PR TITLE
Feature/apl 609 derived tables list

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**
Many changes were merged from 'develop' branch.
Added unit test for retrieving 'derived table data' and using it for later data export/import

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java [hunk 1]
```diff
@@ -40,13 +48,128 @@ public static void close(AutoCloseable... closeables) {
...
+     * Create an output stream to write into the file.
+     * This method is similar to Java 7
+     * <code>java.nio.file.Path.newOutputStream</code>.
+     *
+     * @param fileName the file name
+     * @param append if true, the file will grow, if false, the file will be
+     *            truncated first
+     * @return the output stream
+     */
+    public static OutputStream newOutputStream(String fileName, boolean append)
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java]: Output stream is not Db related, move to another class

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java [hunk 2]
```diff
@@ -40,13 +48,128 @@ public static void close(AutoCloseable... closeables) {
...
+            throws IOException {
+        Objects.requireNonNull(fileName, "fileName is NULL");
+        FilePath filePath = FilePath.get(fileName);
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java]: Do we actually require FilePath from H2?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedDbTable.java [hunk 3]
```diff
@@ -154,8 +154,70 @@ public  DatabaseManager getDatabaseManager() {
...
+        TransactionalDataSource dataSource = databaseManager.getDataSource();
+        try (Connection con = dataSource.getConnection();
+             PreparedStatement pstmt = con.prepareStatement(selectMinSql)) {
+            pstmt.setInt(1, height);
+            try (ResultSet rs = pstmt.executeQuery()) {
+                if (rs.next()) {
+                    dbIdMin = rs.getLong("min_db_id");
+                    dbIdMax = rs.getLong("max_db_id");
+                    long rowCount = rs.getLong("count");
+                    minMaxDbId = new MinMaxDbId(dbIdMin - 1, dbIdMax + 1); // plus/minus one in Max/Min value
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedDbTable.java]: Why -1?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedTableInterface.java [hunk 4]
```diff
@@ -24,4 +30,17 @@
 
     boolean delete(T t);
 
+    default DerivedTableData<T> getAllByDbId(MinMaxDbId minMaxDbId, int limit) throws SQLException {
+        throw new UnsupportedOperationException("GetAll is not supported");
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedTableInterface.java]: These default methods were implemented in DerivedDbTable class, please remove default implementations

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+        super.dataExportPath = Objects.requireNonNull(dataExportPath, "dataExportPath is NULL");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ResultSet read(String inputFileName, String[] colNames, String charset) throws SQLException {
+        Objects.requireNonNull(inputFileName, "inputFileName is NULL");
+        assignNewFileName(inputFileName, true);
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java]: Method do nothing

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+                input = new InputStreamReader(in, characterSet);
+            } catch (IOException e) {
+                close();
+                throw e;
+            }
+        }
+        if (!input.markSupported()) {
+            input = new BufferedReader(input);
+        }
+        input.mark(1);
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java]: Do we require such compatibility?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/jdbc/SimpleResultSet.java [hunk 7]
```diff
@@ -0,0 +1,2467 @@
...
+import java.sql.Statement;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.sql.Types;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Map;
+
+import com.apollocurrency.aplwallet.apl.core.shard.util.ConversionUtils;
+import org.h2.api.JavaObjectSerializer;
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/jdbc/SimpleResultSet.java]: Maybe remove h2 dependencies?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java]: Moved into separed CsvFileUtils.java

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DbUtils.java]: Removed completely. Path is taken from configs.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedDbTable.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedDbTable.java]: That works good on pagination flow as we retrieve lower bound value in on first loop:
`... where db_id > ? and db_id < ? limit ?`

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedTableInterface.java (hunk 4 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/DerivedTableInterface.java]: Removed completely

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java (hunk 5 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java]: All magic logic happens inside - 
`return readResultSet(colNames);`

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java (hunk 6 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/csv/CsvReaderImpl.java]: I would like do not changed that H2 borrowed code now.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/jdbc/SimpleResultSet.java (hunk 7 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/jdbc/SimpleResultSet.java]: I tried, but those three classes are hard to replace quickly.

**Pull request merged**